### PR TITLE
0.12.1 Cherry-Picks: Part 2

### PR DESCRIPTION
**Pull request by kbendick**
This PR is to get the following two PRs, slated for release in 0.12.1, to be able to merge.

Neither of them cherry-pick cleanly, though I do know the set of PRs that would be needed to do so (as mentioned in another PR).
- https://github.com/apache/iceberg/pull/3195
- https://github.com/apache/iceberg/pull/3199

** Merge first - After a decision is made about #3264

### CODE REVIEW

**Comment by kbendick** [comment]
This should be included before https://github.com/apache/iceberg/pull/3428.

This also doesn't need https://github.com/apache/iceberg/pull/3264, but I added it to try to make the test suite run faster yesterday as it was really slow (maybe just noisy neighbors on the runners).

File: core/src/main/java/org/apache/iceberg/BaseOverwriteFiles.java [hunk 1]
```diff
@@ -93,9 +100,22 @@ public OverwriteFiles caseSensitive(boolean isCaseSensitive) {
...
+  public OverwriteFiles validateNoConflictingData() {
+    this.validateNewDataFiles = true;
+    failMissingDeletePaths();
+    return this;
+  }
+
+  @Override
+  public OverwriteFiles validateNoConflictingDeletes() {
+    this.validateNewDeleteFiles = true;
     failMissingDeletePaths();
```
Comment (kbendick) [inline core/src/main/java/org/apache/iceberg/BaseOverwriteFiles.java:119]: The fact that both of these have `failMissingDeletePaths` makes me think something happened with the cherry-picks. Investigating further.

Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/BaseOverwriteFiles.java:119]: Actually, this is the way both of these functions look in master - Is this intentional?:

https://github.com/apache/iceberg/blob/fad77e8b4ee811f6bcbd1c440708e0bc5e1929c2/core/src/main/java/org/apache/iceberg/BaseOverwriteFiles.java#L100-L111

Wondering if this is correct.

Comment (kbendick) [reply core/src/main/java/org/apache/iceberg/BaseOverwriteFiles.java:119]: Nevermind. This looks correct. Here's the actual function called:

https://github.com/apache/iceberg/blob/fad77e8b4ee811f6bcbd1c440708e0bc5e1929c2/core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java#L143-L146

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/BaseOverwriteFiles.java:119]: Yes, this is intentional. This was to preserve the old behavior.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Looks great! I merged these into 0.12.x. Thanks, @kbendick!

**Pull request merged**